src/fuzzing/mutator_test.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/fuzzing/mutator.h"
+
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace fido2_tests {
+namespace mutator {
+namespace {
+
+bool CheckStrictAscending(std::vector<uint8_t> const& data) {
+  if (data.empty()) return true;
+  for (int i = 1; i < data.size(); ++i) {
+    if (data[i] <= data[i - 1]) return false;
+  }
+  return true;
+}
+
+TEST(Mutator, TestEraseByte) {
+  std::vector<uint8_t> data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  std::vector<uint8_t> empty_data = {};
+  ASSERT_FALSE(mutator::EraseByte(empty_data, /* max_size = */ 0));
+  for (int i = 0; i < data.size(); ++i) {
+    srand(i);
+    size_t expected_current_size = data.size() - 1;
+    ASSERT_TRUE(mutator::EraseByte(data, data.size()));
+    ASSERT_TRUE(data.size() == expected_current_size);
+    EXPECT_TRUE(CheckStrictAscending(data));
+  }
+}
+
+TEST(Mutator, TestInsertByte) {
+  std::vector<uint8_t> data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  for (int i = 0; i < 100; ++i) {
+    srand(i);
+    size_t expected_current_size = data.size() + 1;
+    std::multiset<uint8_t> data_before_mutation(data.begin(), data.end());
+    ASSERT_FALSE(mutator::InsertByte(data, /* max_size = */ 1));
+    ASSERT_TRUE(mutator::InsertByte(data, expected_current_size));
+    ASSERT_TRUE(data.size() == expected_current_size);
+    std::multiset<uint8_t> data_after_mutation(data.begin(), data.end());
+    EXPECT_TRUE(std::includes(
+        data_after_mutation.begin(), data_after_mutation.end(),
+        data_before_mutation.begin(), data_before_mutation.end()));
+  }
+}
+
+TEST(Mutator, TestShuffleBytes) {
+  std::vector<uint8_t> data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  for (int i = 0; i < 100; ++i) {
+    srand(i);
+    size_t expected_current_size = data.size();
+    std::multiset<uint8_t> data_before_mutation(data.begin(), data.end());
+    ASSERT_FALSE(mutator::ShuffleBytes(data, /* max_size = */ 1));
+    ASSERT_TRUE(mutator::ShuffleBytes(data, expected_current_size));
+    ASSERT_TRUE(data.size() == expected_current_size);
+    std::multiset<uint8_t> data_after_mutation(data.begin(), data.end());
+    ASSERT_TRUE(data_before_mutation == data_after_mutation);
```

Comment:
Shouldn't this be false?

Reply:
I'm comparing the set here so shuffle should keep the elements set invariant. I added the change check in the new commit

Reply:
Oop thanks!